debuginfo/src/dwarf.rs

```diff
@@ -675,100 +711,136 @@ impl<'d, 'a> DwarfUnit<'d, 'a> {
                 // function. In general, a compiler should always output the call line and call
                 // file for inlined subprograms. If this info is missing, the lookup might
                 // return invalid line numbers.
+                // All the lines have been collected in the parent so just get the lines from the parent
+                // which belong to each range in the inlinee
                 if let (Some(line), Some(file_id)) = (call_line, call_file) {
                     let file = self.resolve_file(file_id).unwrap_or_default();
-                    let lines = &mut parent.lines;
-
-                    // Obtain the index of the first line record in the parent that fully points
-                    // into this inlinee.
-                    let start = match lines.binary_search_by_key(&function_address, |l| l.address) {
-                        Ok(idx) => idx,
-                        Err(idx) => {
-                            // Fix the length of the previous line record to go up to the start of
-                            // the inline function. This effectively splits the previous record in
-                            // two.
-                            if let Some(prev) = idx.checked_sub(1).and_then(|i| lines.get_mut(i)) {
-                                let max_size = function_address - prev.address;
-                                if prev.size.map_or(true, |prev_size| prev_size > max_size) {
-                                    prev.size = Some(max_size);
-                                }
-                            }
+                    let parent_lines = &mut parent.lines;
 
-                            // Insert a new record pointing to the correct call location. Note that
-                            // "base_dir" can be inherited safely here.
-                            let size = match lines.get(idx) {
-                                Some(next) => next.address - function_address,
-                                None => function_size,
-                            };
-
-                            let line_info = LineInfo {
-                                address: function_address,
-                                size: Some(size),
-                                file: file.clone(),
-                                line,
-                            };
-
-                            lines.insert(idx, line_info);
-                            idx + 1
-                        }
-                    };
-
-                    // Patch all following line records pointing them to the call location. Stop at
-                    // the boundary of the inlinee.
-                    for index in start..lines.len() {
-                        let record = &mut lines[index];
-                        if record.address >= function_end {
-                            break;
+                    for range in range_buf.iter() {
```

Comment:
In this loop, we now perform two binary searches per range. Especially for functions with many functions, it may be considerably faster to just iterate sequentially through parent lines.

What if we stick with the initial binary search of the previous implementation, but then we change the for loop (`for index in start..lines.len()` in the original code) to check for membership in the respective range, or otherwise advance the range? This might help to keep the diff smaller.